### PR TITLE
ensure selected cracker is available and viable

### PR DESCRIPTION
**Pull request by jmartin-tech**
When no password cracker is installed `cracker_version` returns `nil`.
Guard against `nil` in the version check and consolidate detection of
a viable environment and application as a responsibility of the factory
that provides the instance.

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole` on a system without `john` or `hashcat` installed and a empty database available
- [ ] `db_disconnect`
- [ ] `use crack_linux` (or any other `auxiliary/analyze/crack_*` module)
- [ ] `run`
- [ ] **Verify** Failure is reported with database requirement noted.
- [ ] Start `msfconsole` on a system without `john` or `hashcat` installed
- [ ] `db_connect`
- [ ] `use crack_linux`
- [ ] `run`
- [ ] **Verify** Failure is reported noting no binary found.
- [ ] install `john` without `jumbo` patch
- [ ] `use crack_linux`
- [ ] `run`
- [ ] **Verify** Failure is reported noting `jumbo` patch is required.
- [ ] install `john` with `jumbo` patch
- [ ] `use crack_linux`
- [ ] `run`
- [ ] **Verify** Result reports no credentials cracked since none exist in the database.
- [ ] `use crack_linux`
- [ ] `set action hashcat`
- [ ] `run`
- [ ] **Verify** Failure is reported noting no binary found.
- [ ] install `hashcat`
- [ ] `use crack_linux`
- [ ] `set action hashcat`
- [ ] `run`
- [ ] **Verify** Result reports no credentials cracked since none exist in the database.

### CODE REVIEW

File: modules/auxiliary/analyze/crack_databases.rb [hunk 1]
```diff
@@ -208,15 +208,9 @@ def check_results(passwords, results, hash_type, hashes, method, cracker)
     # Inner array format: db_id, hash_type, username, password, method_of_crack
     results = []
 
-    cracker = new_password_cracker
+    cracker = new_password_cracker(action.name)
     cracker.cracker = action.name
```
Comment (h00die) [inline modules/auxiliary/analyze/crack_databases.rb:212]: based on all the other changes, i think you missed removing this line

**Review by h00die — COMMENTED** [review]
untested, but that does clean the code up!

File: modules/auxiliary/analyze/crack_databases.rb (hunk 1 above)
Comment (jmartin-r7) [reply modules/auxiliary/analyze/crack_databases.rb:212]: Good catch! Fixed.

File: lib/msf/core/auxiliary/password_cracker.rb [hunk 2]
```diff
@@ -75,16 +75,28 @@ def john_lm_upper_to_ntlm(pwd, hash)
   #
   # @return [nilClass] if there is no active framework db connection
   # @return [Metasploit::Framework::PasswordCracker::Cracker] if it successfully creates a Password Cracker object
-  def new_password_cracker
-    return nil unless framework.db.active
-    Metasploit::Framework::PasswordCracker::Cracker.new(
+  def new_password_cracker(cracking_application)
+    fail_with(Msf::Module::Failure::BadConfig, "Password cracking is not available without an active data database connection.") unless framework.db.active
```
Comment (adfoster-r7) [inline lib/msf/core/auxiliary/password_cracker.rb:79]: I wasn't sure if this was intentional or not:

```suggestion
    fail_with(Msf::Module::Failure::BadConfig, "Password cracking is not available without an active database connection.") unless framework.db.active
```

**Review by adfoster-r7 — APPROVED** [review]
Everything looks good to me; Just want to confirm that open question then I can land this :+1:

File: lib/msf/core/auxiliary/password_cracker.rb (hunk 2 above)
Comment (jmartin-r7) [reply lib/msf/core/auxiliary/password_cracker.rb]: Good catch!... Fixed.

**Comment by adfoster-r7** [comment]
## Release Notes

Improves error message feedback when using the `auxiliary/analyze/crack_*` modules. Examples include notifying the user that the database needs to be active, and having JohnTheRipper Jumbo patch installed

**Pull request merged**